### ISSUE TITLE
PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent failed in CI

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#39780]
```
Alpine.39.Amd64.Open-x64-Debug
Get Repro environment
Unhandled Exception of Type Xunit.Sdk.NullException
Message :
Assert.Null() Failure
Expected: (null)
Actual:   Byte[] [42, 42, 42, 42, 42, ...]
Stack Trace :
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.<>c__DisplayClass67_0.<<PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent>b__1>d.MoveNext() in /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:line 1926
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 83
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 111
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 71
   at System.Net.Test.Common.Http2LoopbackServer.CreateClientAndServerAsync(Func`2 clientFunc, Func`2 serverFunc, Int32 timeout) in /_/src/Common/tests/System/Net/Http/Http2LoopbackServer.cs:line 186
   at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent() in /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:line 1898
--- End of stack trace from previous location where exception was thrown ---
```

cc: @geoffkizer

**Comment by geoffkizer** [comment]
Probably a timing issue, not a product issue. The default Expect: 100-continue timeout is pretty low. I'll change the test to increase it.

**Comment by geoffkizer** [comment]
From the linked PR:

On further inspection, I think there's a real product issue here. The _expect100ContinueWaiter doesn't get created until we call SendRequestBodyAsync. It's possible we could have received the 403 response before this gets created, meaning it will never get signalled.

We need to create the 100-continue waiter in the constructor, because we can receive the 403 any time after this. And we need to not clear it when we fire it on receiving the 403, because then we won't be able to wait on it.

**Comment by geoffkizer** [comment]
We should discuss whether we want to try to get this into 3.0 or not. If not, it should go into a servicing branch.

The impact here is that if you use Expect: 100-continue, and the server sends a 100 Continue or an early failure response, on occasion this will not trigger the request body send, due to the race above. 

Instead, we'll wait for the expect timeout and then send the body. For an early failure response, this means we will send the body instead of cancelling it, but that's always a risk here due to timing.

GRPC does not use Expect: 100-continue.

@karelz @davidsh @stephentoub opinions?

**Comment by stephentoub** [comment]
> We should discuss whether we want to try to get this into 3.0 or not. If not, it should go into a servicing branch.

If it's important enough to be serviced for, we should get it into 3.0.

**Comment by stephentoub** [comment]
Failed again here:
https://dev.azure.com/dnceng/public/_build/results?buildId=283773&view=ms.vss-test-web.build-test-results-tab&runId=7926860&resultId=368519&paneView=attachments
```
    System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2.PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent [FAIL]
      Assert.Null() Failure
      Expected: (null)
      Actual:   Byte[] [42, 42, 42, 42, 42, ...]
      Stack Trace:
        /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(1926,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.<>c__DisplayClass67_0.<<PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent>b__1>d.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
        /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(83,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks)
        /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(111,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks)
        /_/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(71,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout)
        /_/src/Common/tests/System/Net/Http/Http2LoopbackServer.cs(186,0): at System.Net.Test.Common.Http2LoopbackServer.CreateClientAndServerAsync(Func`2 clientFunc, Func`2 serverFunc, Int32 timeout)
        /_/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(1898,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.PostAsyncExpect100Continue_NonSuccessResponse_RequestBodyNotSent()
        --- End of stack trace from previous location where exception was thrown ---
```

**Comment by karelz** [comment]
How much does it impact 3.0? Should we port it?
Is it test only change?

**Comment by wfurt** [comment]
closed by #39932

**Comment by karelz** [comment]
Fixed
